examples/wakaama/lwm2mclient.c

```diff
@@ -0,0 +1,646 @@
+            break;
+         case STATE_READY:
+            fprintf(stdout, "STATE_READY\ren");
+            break;
+         default:
+            fprintf(stdout, "Unknown...\r\n");
+            break;
+      }
+      if (result != 0) {
+         fprintf(stderr, "lwm2m_step() failed: 0x%X\r\n", result);
+#ifdef LWM2M_BOOTSTRAP
+         if (previousState == STATE_BOOTSTRAPPING) {
+#ifdef WITH_LOGS
+            fprintf(stdout,
+                    "[BOOTSTRAP] restore security and server objects\r\n");
+#endif /* WITH_LOGS */
+            prv_restore_objects(lwm2mH);
+            lwm2mH->state = STATE_INITIAL;
+         } else
+#endif /* LWM2M_BOOTSTRAP */
+         {
+            g_quit = 1;
+            continue;
+         }
+      }
+#ifdef LWM2M_BOOTSTRAP
+      DEBUG("update_bootstrap_info()\r\n");
+      update_bootstrap_info(&previousState, lwm2mH);
+#endif
+
+   handshake:
+      DEBUG("xtimer_msg_receive_timeout64()\r\n");
+      //if (xtimer_msg_receive_timeout64(&msg, tv.tv_sec * 1000000) >= 0) {
+      pckt_rcvd_size = sock_udp_recv(&data.sock, &packet_rcvd, DTLS_MAX_BUF, tv.tv_sec * US_PER_SEC, &remote);
+      if (pckt_rcvd_size > 0) {
+         connP = connection_find(data.connList, &remote);
+         if (connP != NULL) {
+            DEBUG("connection_handle_packet(%d)\r\n", pckt_rcvd_size);
+            int result = connection_handle_packet(connP, packet_rcvd, pckt_rcvd_size);
+            if (0 != result) {
+               printf("error handling message %d\n", result);
+            }
+
+            /*
+             * We loop around here until the DTLS handshake is done,
+             * since lwm2m_step() will try to send stuff before then and fail
+             * without retrying.
+             */
+            if (!connP->noSec) {
+               dtls_peer_t *peer =
+                   dtls_get_peer(connP->dtlsContext, connP->dtlsSession);
+               if (peer->state != DTLS_STATE_CONNECTED) {
+                  goto handshake;
+               }
+            }
+         }
+      } else if (pckt_rcvd_size == -ETIMEDOUT) {
+         // Check for DTLS handshake timeout of the last connection
+
+         dtls_connection_t *c;
+
+         for (c = data.connList; c != NULL; c = c->next) {
+            if (!c->noSec) {
+               DEBUG("Checking connections for timeouts\n");
+               dtls_peer_t *peer = dtls_get_peer(c->dtlsContext, c->dtlsSession);
+               if (peer->state != DTLS_STATE_CONNECTED) {
+                  dtls_reset_peer(c->dtlsContext, peer);
+                  fprintf(stderr, "Error: DTLS handshake timed out\n");
+                  if (lwm2mH->state != STATE_READY) {
+                     DEBUG("Quitting\n");
+                     g_quit = 1;
+                  }
+               }
+            }
+         }
+      }
+      // TODO: Error handling for sock_udp_recv()
+   }
+
+   /*
+    * Finally when the loop is left smoothly - asked by user in the command line
```

Comment:
I still need to read your code a little more (and the package). 

But, are we sure we want to terminate this thread? Is this one who has all DTLS context and even Wakama data.

Also, you are not adding any special to the shell command list.

Reply:
The `exit` command is added in `main.c`.

And as the main loop can only be cleanly termianted by changing the value of `g_quit`, which currently only happens in the SIGINT signal handler. This was basically just meant to make the program quitable on native (which still doesn't work properly yet, hence the `exit` shell command).